spacy/tests/doc/test_underscore.py

```diff
@@ -171,3 +171,343 @@ def test_method(doc, arg1=1, arg2=2):
     doc = Doc(en_vocab, words=["hello", "world"])
     assert test_method.__doc__ == "I am a docstring"
     assert doc._.test_docstrings.__doc__.rsplit(". ")[-1] == "I am a docstring"
+
+
+def test_underscore_for_unique_span(en_tokenizer):
+    """Test that spans with the same boundaries but with different labels are uniquely identified (see #9706)."""
+    Doc.set_extension(name="doc_extension", default=None)
+    Span.set_extension(name="span_extension", default=None)
+    Token.set_extension(name="token_extension", default=None)
+
+    # Initialize doc
+    text = "Hello, world!"
+    doc = en_tokenizer(text)
+    span_1 = Span(doc, 0, 2, "SPAN_1")
+    span_2 = Span(doc, 0, 2, "SPAN_2")
+
+    # Set custom extensions
+    doc._.doc_extension = "doc extension"
+    doc[0]._.token_extension = "token extension"
+    span_1._.span_extension = "span_1 extension"
+    span_2._.span_extension = "span_2 extension"
+
+    # Assert extensions
+    assert (
+        doc.user_data[
+            (
+                "._.",
+                "span_extension",
+                span_1.start_char,
+                span_1.end_char,
+                span_1.label,
+                span_1.kb_id,
+                span_1.id,
+            )
+        ]
+        == "span_1 extension"
+    )
```

Comment:
To make these tests a bit more compact & readible, I suggest we introduce a helper function in this file:
```
def _get_tuple(s: Span):
    return "._.", "span_extension", s.start_char, s.end_char, s.label, s.kb_id, s.id
```
after which all `assert` statements can be just one line:
```
assert doc.user_data[_get_tuple(span_1)] == "span_1 extension"
```
This will also be useful for the other unit tests.